api/board/webidl.md

```diff
@@ -0,0 +1,179 @@
+  unsigned long rate;
+};
+
+[Constructor( (PinName or AIOOptions) pin, optional Board board)]
+interface AIO: Pin {
+    unsigned long channel;  // analog channel
+    unsigned long rateLimit;     // rate limit for ondata
+    Promise<unsigned long> read();  // one-shot async read
+    void close();
+    attribute EventHandler<unsigned long> ondata;
+};
+
+AIO implements EventEmitter;
+
+// PWM
+dictionary PWMOptions {
+  PinName pin;
+  boolean reversePolarity = false;
+};
+
+[Constructor(PWMOptions options, optional Board board)]
+interface PWM: Pin {
+    readonly attribute unsigned long channel;
+    readonly attribute boolean reversePolarity;
+    // 'value' returns PWMData
+    void write(PWMData value);
+    void stop();
+    void close();
+};
+
+dictionary PWMData {
+  double period;
+  double pulseWidth;
+  double dutyCycle;
+};
+
+// I2C
+enum I2CSpeed { "10kbps", "100kbps", "400kbps", "1000kbps", "3400kbps" };
+
+dictionary I2COptions {
+  octet bus;
+  I2CSpeed speed;
+};
+
+[NoInterfaceObject]
+interface I2C {
+  readonly attribute octet bus;
+  readonly attribute I2CBusSpeed speed;
+  readonly attribute boolean busy;
+
+  Promise init(I2COptions options);
+  void requestRead(octet device,
+                                unsigned long size,
+                                optional octet register,
+                                optional repetitions = 1);
+  Promise write(octet device, (USVString or sequence<octet>) data, optional octet register);
+  Promise writeBit(octet device, boolean data);
+  void abort();  // abort all current read / write operations
+  void close();
+
+  attribute EventHandler<sequence<octet>> ondata;
+  attribute EventHandler onerror;
+};
+
+I2C implements EventEmitter;
+
+// SPI
+typedef (sequence<octet> or ArrayBuffer) SPIData;
+
+enum SPIMode {
+
+  "mode0",  // polarity normal, phase 0, i.e. sampled on leading clock
+  "mode1",  // polarity normal, phase 1, i.e. sampled on trailing clock
+  "mode2",  // polarity inverse, phase 0, i.e. sampled on leading clock
+  "mode3"   // polarity inverse, phase 1, i.e. sampled on trailing clock
+};
+
+enum SPIDataOrder { "msb", "lsb" };
+
+dictionary SPIOptions {
+  unsigned long bus = 0;
+  SPIMode mode = "mode0";
+  boolean msb = 1;  // 1: MSB first, 0: LSB first
+  unsigned long dataBits = 8; // 1, 2, 4, 8, 16
+  double speed = 20;  // in MHz, usually 10..66 MHz
+};
+
+[NoInterfaceObject]
+interface SPI {
+  // has all the properties of SPIOptions as read-only attributes
+  Promise init(SPIOptions options);
+  Promise<SPIData> transfer(SPIData txData);
+  void close();
+};
+
+// UART
+enum UARTBaud { "baud-9600", "baud-19200", "baud-38400", "baud-57600", "baud-115200" };
+enum UARTDataBits { "databits-5", "databits-6", "databits-7", "databits-8" };
+enum UARTStopBits { "stopbits-1", "stopbits-2" };
+enum UARTParity { "none", "even", "odd" };
+
+dictionary UARTOptions {
+  DOMString port;
+  UARTBaud baud = "115200";
+  UARTDataBits dataBits = "8";
```

Comment:
I rather have these as numbers... btw with your code this should not be "8" but "databits-8". but just make it 8 and bail out with wrong numbers